### PR TITLE
Validation fix

### PR DESCRIPTION
**Pull request by morsecodist**
# Fixed a few issues:

## Removed Depreciated Calls

The repository had some depreciated calls: `Dict([x => y])` rather than `Dict(x => y)` and `nothing` rather than `Void`.

## Addressed Julia Warnings

Every time I loaded the package I got an error that the following operator should be explicitly imported from base: 

- `.*`
- `*`
- `/`
- `.-`
- `-`
- `.+`
- `+`
- `setindex!`

so I explicitly imported them. 

## Fixed logging issue

The `diagnostic_trace!` function was only printing validation error when `valid` was false. I changed it to print validation error when `valid` was true.

# Added validation set functionality

## Added validation set to adatrain

Handled validation sets in `adatrain` the same way they were already handled in `gdmtrain`.

## Added check for convergence with validation set

If the validation set is present `e_valid_old` and `e_valid` are updated each iteration and used to check for convergence.

### CODE REVIEW

**Comment by anj1** [comment]
Hi, if you want to re-open it I can merge it.

Just to let you know though, I consider this library to be obsolete and I am currently using TensorFlow for my own work.

**Comment by morsecodist** [comment]
I mistakenly committed using the wrong author identity so I had to change the author of the commits and submit a new pull request. Sorry about that, thanks for the merge. I know it's obsolete but it is my first open source contribution and it seemed pretty approachable.
